File: library/src/main/java/world/rafoufoun/providerdelegate/DelegationProvider.java
package world.rafoufoun.providerdelegate;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

/**
 * A ready to go {@link ContentProvider} implementation having a {@link ProviderDelegateManager}
 * ready to use. You can call {@link ProviderDelegateManager#addDelegate(ProviderDelegate)}
 * at any time if you have called {@link #onCreate()} into your subclass {@link DelegationProvider#onCreate()}
 */
public abstract class DelegationProvider extends ContentProvider {

    private ProviderDelegateManager delegateManager;

    /**
     * Initialize the ProviderDelegateManager.
     *
     * @return return true if the initialization is ok.
     */
    @Override
    public final boolean onCreate() {
        delegateManager = new ProviderDelegateManager(getContext(), getAuthority());
        initializeProvider();
        addProviderDelegate(delegateManager);
        return true;
    }

    /**
     * Method called from {@link DelegationProvider#onCreate()}
     * This stand for initializing anything you need that would normally be in {@link ContentProvider#onCreate()}
     */
    public abstract void initializeProvider();

    /**
     * Thi method is called from {@link DelegationProvider#onCreate()}
     * Add your {@link ProviderDelegate} here with {@link ProviderDelegateManager#addDelegate(ProviderDelegate)}
     */
    protected abstract void addProviderDelegate(ProviderDelegateManager delegateManager);

    /**
     * Get the authority for this provider
     *
     * @return return the provider authority
     */
    protected abstract String getAuthority();

    protected abstract SQLiteDatabase getWritableDatabase();

    protected abstract SQLiteDatabase getReadableDatabase();

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return delegateManager.getType(uri);
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        return delegateManager.query(getReadableDatabase(), uri, projection, selection, selectionArgs, sortOrder);
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, ContentValues values) {
        return delegateManager.insert(getWritableDatabase(), uri, values);
    }

    @Override
    public int delete(@NonNull Uri uri, String selection, String[] selectionArgs) {
        return delegateManager.delete(getWritableDatabase(), uri, selection, selectionArgs);
    }

    @Override
    public int update(@NonNull Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        return delegateManager.update(getWritableDatabase(), uri, values, selection, selectionArgs);
    }
}

File: library/src/main/java/world/rafoufoun/providerdelegate/DelegationProvider.kt
package world.rafoufoun.providerdelegate

import android.content.ContentProvider
import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.net.Uri

/**
 * A ready to go [ContentProvider] implementation having a [ProviderDelegateManager]
 * ready to use. You can call [ProviderDelegateManager.addDelegate]
 * at any time if you have called [.onCreate] into your subclass [DelegationProvider.onCreate]
 */
abstract class DelegationProvider : ContentProvider() {

    private lateinit var delegateManager: ProviderDelegateManager

    /**
     * Get the authority for this provider
     *
     * @return return the provider authority
     */
    protected abstract val authority: String

    protected abstract val writableDatabase: SQLiteDatabase

    protected abstract val readableDatabase: SQLiteDatabase

    /**
     * Initialize the ProviderDelegateManager.
     *
     * @return return true if the initialization is ok.
     */
    override fun onCreate(): Boolean {
        delegateManager = ProviderDelegateManager(context!!.contentResolver, authority)
        initializeProvider()
        addProviderDelegate(delegateManager)
        return true
    }

    /**
     * Method called from [DelegationProvider.onCreate]
     * This stand for initializing anything you need that would normally be in [ContentProvider.onCreate]
     */
    abstract fun initializeProvider()

    /**
     * Thi method is called from [DelegationProvider.onCreate]
     * Add your [ProviderDelegate] here with [ProviderDelegateManager.addDelegate]
     */
    protected abstract fun addProviderDelegate(delegateManager: ProviderDelegateManager)

    override fun getType(uri: Uri): String {
        return delegateManager.getType(uri)
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor {
        return delegateManager.query(readableDatabase, uri, projection, selection, selectionArgs, sortOrder)
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri {
        return delegateManager.insert(writableDatabase, uri, values)
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        return delegateManager.delete(writableDatabase, uri, selection, selectionArgs)
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        return delegateManager.update(writableDatabase, uri, values, selection, selectionArgs)
    }
}
